packages/blocks/src/types.ts

```diff
@@ -0,0 +1,976 @@
+
+export type TransformRawSchema = {
+	[ k in keyof HTMLElementTagNameMap | '#text' ]?: {
+		attributes?: string[] | undefined;
+		require?: Array< keyof HTMLElementTagNameMap > | undefined;
+		classes?: Array< string | RegExp > | undefined;
+		children?: TransformRawSchema | undefined;
+		[ key: string ]: any; // Permit untyped/undocumented property access.
+	};
+};
+
+export interface TransformBlock< T extends Record< string, any > > {
+	type: 'block';
+	priority?: number | undefined;
+	blocks: string[];
+	isMatch?( attributes: T, block: string | string[] ): boolean;
+	isMultiBlock?: boolean | undefined;
+	transform( attributes: T ): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformEnter< T extends Record< string, any > > {
+	type: 'enter';
+	priority?: number | undefined;
+	regExp: RegExp;
+	transform(): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformFiles< T extends Record< string, any > > {
+	type: 'files';
+	priority?: number | undefined;
+	isMatch?( files: FileList ): boolean;
+	transform(
+		files: FileList,
+		onChange?: ( id: string, attrs: T ) => void
+	): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformPrefix< T extends Record< string, any > > {
+	type: 'prefix';
+	priority?: number | undefined;
+	prefix: string;
+	transform( content: string ): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformRaw< T extends Record< string, any > > {
+	type: 'raw';
+	priority?: number | undefined;
+	/**
+	 * Comma-separated list of selectors, no spaces.
+	 *
+	 * @example 'p,div,h1,.css-class,#id'
+	 */
+	selector?: string | undefined;
+	schema?: TransformRawSchema | undefined;
+	isMatch?( node: Node ): boolean;
+	transform?( node: Node ): WPBlock< Partial< T > > | void;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformShortcode< T extends Record< string, any > > {
+	type: 'shortcode';
+	priority?: number | undefined;
+	tag: string;
+	transform?(
+		attributes: any,
+		match: Record< string, any > // TODO type shortcode
+	): WPBlock< T >;
+	attributes?: any; // TODO find type
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export type Transform<
+	T extends Record< string, any > = Record< string, any >
+> =
+	| TransformBlock< T >
+	| TransformEnter< T >
+	| TransformFiles< T >
+	| TransformPrefix< T >
+	| TransformRaw< T >
+	| TransformShortcode< T >;
+
+export type WPBlockAttributes = Record< string, any >;
+
+/**
+ * TODO undocumented
+ */
+export type InnerBlockTemplate = [
+	string,
+	WPBlockAttributes?,
+	InnerBlockTemplate[]?
+];
+
+export type WPBlockVariationScope = 'block' | 'inserter' | 'transform';
+
+export interface WPBlockVariation<
```

Comment:
I noticed some types are prefixed with `WP` and some aren't. Is there a need for this prefix? It's quite clear this is a part of WordPress.

Reply:
I initially created the types without prefix (as done in `@types/wordpress__blocks`), but then found out that other package mention some of them them with prefix and renaming the core types was beyond the scope of what I am trying to accomplish.

I would also prefer un-prefixed, name collision isn't a concern in modules. But I figured it wasn't my decision to make. I agree they should be consistent, if there is agreement on a pattern to follow I can make the changes.

Reply:
Strong vote in favor of moving away from the `WP` prefix. It doesn't add anything.

Reply:
I removed the `WP` prefix from the types of this packages. We will need to gage the impact this has throughout the documentation. Nothings should break, `@wordpress/blocks` didn't export types before, so no packages could have relied on them. But there is documentation else where in the project that references some of the key interfaces of this package with the `WP` prefix.

Reply:
we might want to create a discussion on this for the project. I've long been for removing the WP syntax, but some people still preferred them the last time I brought it up.

Reply:
In the grand scheme of things what their named isn't the hard part, finding the right definitions and where they are used is. So I'll follow whatever the consensus is, and renaming is easy.

Reply:
@dmsnell Where is the best place to start a discussion on a types naming convention?

Reply:
https://github.com/WordPress/gutenberg/discussions/categories/developer-experience?discussions_q=category%3A%22Developer+Experience%22+types